src/pretix/base/datasync/datasync.py

```diff
@@ -0,0 +1,425 @@
+        for (target, event), queued_orders in grouped:
+            target_cls, meta = sync_targets.get(identifier=target, active_in=event)
+            # TODO: what should i do if the sync plugin got deactivated in the meantime?
+            sync_event_to_target(event, target_cls, queued_orders)
+
+
+class SyncConfigError(Exception):
+    def __init__(self, messages, full_message=None):
+        self.messages = messages
+        self.full_message = full_message
+
+
+StaticMapping = namedtuple('StaticMapping', ('pk', 'pretix_model', 'external_object_type', 'pretix_pk', 'external_pk', 'property_mapping'))
+
+
+class OutboundSyncProvider:
+    max_attempts = 5
+
+    def __init__(self, event):
+        self.event = event
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @classmethod
+    @property
+    def display_name(cls):
+        return str(cls.identifier)
+
+    """
+    Adds an order to the sync queue. May only be called on derived classes which define an "identifier" attribute.
+
+    Should be called in the appropriate signal receivers, e.g.:
+
+        @receiver(order_placed, dispatch_uid="mysync_order_placed")
+        def on_order_placed(sender, order, **kwargs):
+            MySyncProvider.enqueue_order(order, "order_placed")
+    """
+    @classmethod
+    def enqueue_order(cls, order, triggered_by, not_before=None):
+        if not hasattr(cls, 'identifier'):
+            raise TypeError('Call this method on a derived class that defines an "identifier" attribute.')
+        OrderSyncQueue.objects.create(
+            order=order,
+            sync_provider=cls.identifier,
+            triggered_by=triggered_by,
+            not_before=not_before)
+
+    @classmethod
+    def get_external_link_info(cls, event, external_link_href, external_link_display_name):
+        return {
+            "href": external_link_href,
+            "val": external_link_display_name,
+        }
+
+    @classmethod
+    def get_external_link_html(cls, event, external_link_href, external_link_display_name):
+        info = cls.get_external_link_info(event, external_link_href, external_link_display_name)
+        return make_link(info, '{val}')
+
+    """
+    Optionally override to configure a different retry backoff behavior
+    """
+    def next_retry_date(self, sq):
+        return now() + timedelta(hours=1)
+
+    """
+    Optionally override this method to exclude certain orders from sync by returning False
+    """
+    def order_valid_for_sync(self, order):
+        return True
+
+    """
+    Implementations must override this property to provide the data mappings as a list of objects.
+
+    They can return instances of the StaticMapping namedtuple defined above, or create their own
+    class (e.g. a Django model).
+
+    The returned objects must have at least the following properties:
+    - pk:                   unique identifier
```

Comment:
"pk" (at least to me) is pretty strongly associated with its meaning "primary key" and, therefore, databases. I found it a little confusing what is expected here. Is there a requirement on the type this has? Unique in what scope?

Reply:
It's called `pk` here because it's somewhat expected that a OutboundSyncProvider implementation will store these mappings in the database (as HubspotSyncer does). Calling it pk allows us to just return the Django models from mappings(). The StaticMapping namedtuple is a helper for less advanced implementations where we just support one or two predefined mappings (as BrevoSyncer does, with one mapping for Orders and one for OrderPositions)